flink-libraries/flink-table/src/test/scala/org/apache/flink/table/formats/TableFormatFactoryServiceTest.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.formats
+
+import java.util.{HashMap => JHashMap, Map => JMap}
+
+import org.apache.flink.table.api.{AmbiguousTableFormatException, NoMatchingTableFormatException}
+import org.apache.flink.table.formats.utils.TestAmbiguousTableFormatFactory
+import org.junit.Assert.{assertNotNull, assertTrue}
+import org.junit.Test
+
+/**
+  * Tests for [[TableFormatFactoryService]].
+  */
+class TableFormatFactoryServiceTest {
+
+  @Test
+  def testValidProperties(): Unit = {
+    val props = properties()
+    assertNotNull(
+      TableFormatFactoryService.find(classOf[TableFormatFactory[_]], props))
+  }
+
+  @Test
+  def testDifferentContextVersion(): Unit = {
+    val props = properties()
+    props.put("format.property-version", "2")
+    // the format should still be found
+    assertNotNull(TableFormatFactoryService.find(classOf[TableFormatFactory[_]], props))
+  }
+
+  @Test
+  def testAmbiguousMoreSupportSelection(): Unit = {
+    val props = properties()
+    props.remove("format.important")
+    props.put("format.special_path", "/what/ever")
+    assertTrue(
+      TableFormatFactoryService
+        .find(classOf[TableFormatFactory[_]], props)
+        .isInstanceOf[TestAmbiguousTableFormatFactory])
+  }
+
+  @Test
+  def testAmbiguousClassBasedSelection(): Unit = {
+    val props = properties()
+    props.remove("format.important")
+    assertTrue(
+      TableFormatFactoryService
+        .find(classOf[TestAmbiguousTableFormatFactory], props)
+        .isInstanceOf[TestAmbiguousTableFormatFactory])
+  }
+
+  @Test
+  def testAmbiguousSchemaBasedSelection(): Unit = {
+    val props = properties()
+    props.remove("format.important")
+    props.put("schema.weird_field", "unknown") // this is unknown to the schema derivation factory
```

Comment:
The format with schema derivation feels not responsible because of this field. But since there is another format that feels responsible no exception is thrown. If `TestAmbiguousTableFormatFactory` would not exist an exception would have been thrown.